docs/numba.md

```diff
@@ -0,0 +1,151 @@
+{meth}`numba_tree_sequence`. This must be done outside `njit` code:
+
+```{code-cell} python
+import msprime
+
+ts = msprime.sim_ancestry(
+    samples=50000,
+    sequence_length=100000,
+    recombination_rate=0.1,
+    random_seed=42
+)
+numba_ts = tskit_numba.numba_tree_sequence(ts)
+print(type(numba_ts))
+```
+
+## Tree Iteration
+
+Tree iteration can be performed using the {class}`NumbaTreeIndex` class.
+This class provides `next()` and `prev()` methods for forward and backward iteration through the trees in a tree sequence. Its `in_range` and `out_range` attributes provide the edges that must be added or removed to form the current
+tree from the previous tree, along with the current tree `interval` and its sites and mutations through `site_range` and `mutation_range`.
+
+A `NumbaTreeIndex` instance can be obtained from a `NumbaTreeSequence` using the `tree_index()` method. The initial state of this is of a "null" tree outside the range of the tree sequence, the first call to `next()` or `prev()`will be to the first, or last tree sequence tree respectively. After that, the `in_range` and `out_range` attributes will provide the edges that must be added or removed to form the current tree from the previous tree. For example
+`tree_index.in_range.order[in_range.start:in_range.stop]` will give the edge ids that are new in the current tree, and `tree_index.out_range.order[out_range.start:out_range.stop]` will give the edge ids that are no longer present in the current tree. `tree_index.site_range` and
+`tree_index.mutation_range` give the indexes into the tree sequences site and mutation arrays.
+
+As a simple example we can calulate the number of edges in each tree in a tree sequence:
+
+```{code-cell} python
+import numba
+
+@numba.njit
+def edges_per_tree(numba_ts):
+    tree_index = numba_ts.tree_index()
+    current_num_edges = 0
+    num_edges = []
+    
+    # Move forward through the trees
+    while tree_index.next():
+        # Access current tree information
+        in_range = tree_index.in_range
+        out_range = tree_index.out_range
+        
+        current_num_edges -= (out_range.stop - out_range.start)
+        current_num_edges += (in_range.stop - in_range.start)
+        num_edges.append(current_num_edges)
+    return num_edges
+```
+
+```{code-cell} python
+:tags: [hide-cell]
+# Warm up the JIT compiler
+edges = edges_per_tree(numba_ts)
+```
+
+
+```{code-cell} python
+import time
+
+t = time.time()
+jit_num_edges = edges_per_tree(numba_ts)
+print(f"JIT Time taken: {time.time() - t:.4f} seconds")
+```
+
+Doing the same thing with the normal `tskit` API would be much slower:
+
+```{code-cell} python
+t = time.time()
+python_num_edges = []
+for tree in ts.trees():
+    python_num_edges.append(tree.num_edges)
+print(f"Normal Time taken: {time.time() - t:.4f} seconds")
+
+assert jit_num_edges == python_num_edges, "JIT and normal results do not match!"
+```
+
```

Comment:
I think we want to include the ``diversity`` calculation here in the documentation - we want to have a real example that people can use as a starting point.

Reply:
Done in https://github.com/tskit-dev/tskit/pull/3225/commits/a17260684236a6a466c0c5a3c0150ef348578719